tests/test_checkout.py

```diff
@@ -68,6 +68,118 @@ def test_merge_diverged_conflict(self):
+            zeroth_co.datasets['_dset']['1']
+
+        assert second_co.metadata['a1'] == 'b1'
+        assert first_co.metadata['a1'] == 'b1'
+        assert first_co.metadata['a2'] == 'b2'
+        with pytest.raises(KeyError):
+            second_co.metadata['a2']
+        with pytest.raises(KeyError):
+            second_co.metadata.get('a2')
+
+    def test_remove_and_add_dset(self, written_repo, array5by7, get_history):
+        co = written_repo.checkout(write=True)
+        co.datasets['_dset']['1'] = array5by7
+        co.datasets['_dset']['2'] = array5by7
+        co.metadata.add('a1', 'b1')
+        co.commit()
+        co.close()
+        co = written_repo.checkout(write=True)
+        co.datasets.remove_dset('_dset')
+        co.commit()
+        co.close()
+        co = written_repo.checkout(write=True)
+        dset = co.datasets.init_dataset('_dset', prototype=array5by7)
+        dset['1'] = np.zeros_like(array5by7)
+        co.commit()
+        history = get_history(co._refenv, co._branchenv)
+        co.close()
+        second_commit = history['order'][2]
+        third_commit = history['order'][1]
+        fourth_commit = history['order'][0]
+        second_co = written_repo.checkout(commit=second_commit)
+        third_co = written_repo.checkout(commit=third_commit)
+        fourth_co = written_repo.checkout(commit=fourth_commit)
+        assert list(second_co.datasets['_dset'].keys()) == ['1', '2']
+        assert np.allclose(second_co.datasets['_dset']['1'], array5by7)
+        with pytest.raises(KeyError):
+            third_co.datasets['_dset']
+        assert np.allclose(fourth_co.datasets['_dset']['1'], np.zeros_like(array5by7))
+
+    def test_3way_merges_with_mutation(self, written_repo, array5by7, get_history):
+        branch1 = written_repo.create_branch('testbranch1')
+        co = written_repo.checkout(write=True)
+        co.datasets['_dset']['1'] = array5by7
+        first_commit = co.commit()
+        co.close()
+        branch2 = written_repo.create_branch('testbranch2')
+        co = written_repo.checkout(write=True, branch_name=branch1)
+        co.datasets['_dset']['2'] = np.ones_like(array5by7) * 4
+        second_commit = co.commit()
+        co.close()
+        branch3 = written_repo.create_branch('testbranch3')
+        co = written_repo.checkout(write=True, branch_name=branch3)
+        tmparray = co.datasets['_dset']['2']
+        tmparray[3] *= 2
+        co.datasets['_dset']['2'] = tmparray
+        third_commit = co.commit()
+        co.close()
+        co = written_repo.checkout(write=True, branch_name=branch2)
+        co.datasets['_dset'].remove('1')
+        fourth_commit = co.commit()
+        co.close()
+        merge_commit1 = written_repo.merge('zeroth and first', 'master', branch1)
+        merge_commit2 = written_repo.merge('data deletion branch', 'master', branch2)
+        # TODO: This is sort of a merge conflict. The error should say something else
```

Comment:
Ok. we should probably just use ValueError's or something for any merge conflicts. please open an issue so we can make the change and reference it here.